### PR TITLE
card: refactor convert_carControl for direct recursive initialization from Cap'n Struct

### PR DESCRIPTION
**Pull request by deanlee**
This PR refactors the `convert_carControl` function to directly and recursively initialize the `CarControl` dataclass from the Cap'n Proto structure, eliminating the need for dictionary conversion and manual assignments. This results in more efficient and maintainable code.

**Changes:**

1. **Direct Recursive Initialization:** The function now recursively initializes dataclass fields directly from the DynamicStructReader,  eliminating the need for the costly `to_dict` conversion.
2. **Field Type Caching:** Introduced a `_get_fieldtypes` method, cached to optimize the retrieval of type hints, reducing redundant type lookups.
3. **Optimized Field Handling:** The new approach ensures that only the necessary fields are processed, with checks in place to handle nested dataclass structures efficiently.
4. **Deprecated Field Removal:** The previous manual removal of deprecated fields is replaced with a more dynamic approach that automatically handles field initialization based on the current dataclass structure.

### CODE REVIEW

File: selfdrive/car/card.py
Comment (sshane) [inline selfdrive/car/card.py:1]: Can we make a file for these functions?

File: selfdrive/car/card.py [hunk 1]
```diff
@@ -111,19 +112,22 @@ def convert_to_capnp(struct: structs.CarParams | structs.CarState | structs.CarC
...
+def convert_carControl(struct: capnp.lib.capnp._DynamicStructReader) -> structs.CarControl:
 
-  struct_dataclass.actuators = structs.CarControl.Actuators(**remove_deprecated(struct_dict.get('actuators', {})))
-  struct_dataclass.cruiseControl = structs.CarControl.CruiseControl(**remove_deprecated(struct_dict.get('cruiseControl', {})))
-  struct_dataclass.hudControl = structs.CarControl.HUDControl(**remove_deprecated(struct_dict.get('hudControl', {})))
+  def initialize_dataclass(cls: Any, data: capnp.lib.capnp._DynamicStructReader) -> Any:
+    fieldtypes = _get_fieldtypes(cls)
+    return cls(**{
+        f: initialize_dataclass(fieldtypes[f], getattr(data, f))
+        if hasattr(fieldtypes[f], _FIELDS) else getattr(data, f)
```
Comment (sshane) [inline selfdrive/car/card.py:126]: use `is_dataclasss`?

**Comment by deanlee** [comment]
Completed. Next, I plan to continue improving the convert_to_capnp function

**Comment by maxime-desroches** [comment]
This PR was closed because of a git history rewrite.
Please read https://github.com/commaai/openpilot/issues/33399 for what to do with your fork and your PRs.

**Pull request closed without merging**